src/main/java/com/michelin/ns4kafka/controller/SchemaController.java

```diff
@@ -160,24 +163,59 @@ public Mono<HttpResponse<Void>> delete(String namespace, @PathVariable String su
             return Mono.error(new ResourceValidationException(SCHEMA, subject, invalidOwner(subject)));
         }
 
-        return schemaService.getLatestSubject(ns, subject)
+        // delete all versions of the schema
+        if (version.isEmpty()) {
+            return schemaService.getLatestSubject(ns, subject)
+                .map(Optional::of)
+                .defaultIfEmpty(Optional.empty())
+                .flatMap(latestSubjectOptional -> {
+                    if (latestSubjectOptional.isEmpty()) {
+                        return Mono.just(HttpResponse.notFound());
+                    }
+
+                    if (dryrun) {
+                        return Mono.just(HttpResponse.noContent());
+                    }
+
+                    Schema schemaToDelete = latestSubjectOptional.get();
+                    sendEventLog(schemaToDelete, ApplyStatus.deleted, schemaToDelete.getSpec(), null);
+
+                    return schemaService
+                        .deleteAllVersions(ns, subject)
+                        .map(deletedVersionIds -> HttpResponse.noContent());
+                });
+        }
+
+        // delete a specific version of the schema
+        return schemaService.getSubject(ns, subject, version.get())
             .map(Optional::of)
             .defaultIfEmpty(Optional.empty())
-            .flatMap(latestSubjectOptional -> {
-                if (latestSubjectOptional.isEmpty()) {
+            .flatMap(subjectOptional -> {
+                if (subjectOptional.isEmpty()) {
                     return Mono.just(HttpResponse.notFound());
                 }
 
                 if (dryrun) {
                     return Mono.just(HttpResponse.noContent());
                 }
 
-                Schema schemaToDelete = latestSubjectOptional.get();
-                sendEventLog(schemaToDelete, ApplyStatus.deleted, schemaToDelete.getSpec(), null);
-
+                Schema schemaToDelete = subjectOptional.get();
                 return schemaService
-                    .delete(ns, subject)
-                    .map(deletedSchemaIds -> HttpResponse.noContent());
+                    .deleteVersion(ns, subject, version.get())
+                    .flatMap(optionalNewLatestSubject -> {
+                        // there was only one version & it was deleted, so the schema is deleted in ns4
+                        if (optionalNewLatestSubject.isEmpty()) {
+                            sendEventLog(schemaToDelete, ApplyStatus.deleted,
+                                schemaToDelete.getSpec(), null);
+                        }
+                        // there were multiple versions & latest version was deleted, so the schema is changed in ns4
+                        if (optionalNewLatestSubject.isPresent() && optionalNewLatestSubject.get()
+                            .getSpec().getVersion() < schemaToDelete.getSpec().getVersion()) {
+                            sendEventLog(schemaToDelete, ApplyStatus.changed,
```

Comment:
Always send a `deleted` status, maybe specifying the version additionally

Reply:
OK, just with the `deleted` status, I refactored the controller method.

Reply:
So I updated the sendEventLog method to add a `version` argument, so I updated the method sendEventLog in many other places too